sunscreen/src/types/zkp/rns_polynomial.rs

```diff
@@ -134,15 +142,61 @@ impl<F: BackendField, const N: usize, const R: usize> MulVar for RnsRingPolynomi
     }
 }
 
+/**
+ * For scaling an algebraic structure (e.g. polynomial.)
+ */
+pub trait Scale<F: BackendField> {
+    /**
+     * Return a structure scaled by `x`.
+     */
+    fn scale(self, x: ProgramNode<NativeField<F>>) -> Self;
+}
+
+impl<F: BackendField, const D: usize, const R: usize> Scale<F>
+    for ProgramNode<RnsRingPolynomial<F, D, R>>
+{
+    fn scale(self, x: ProgramNode<NativeField<F>>) -> Self {
+        let mut output = vec![NodeIndex::from(0); D * R];
+
+        with_zkp_ctx(|ctx| {
+            for (i, o) in output.iter_mut().enumerate().take(R * D) {
+                *o = ctx.add_multiplication(self.ids[i], x.ids[0]);
+            }
+        });
+
+        Self::new(&output)
+    }
+}
+
+impl<F: BackendField, const D: usize, const R: usize> Mod<F> for RnsRingPolynomial<F, D, R> {
+    fn signed_reduce(
+        lhs: ProgramNode<Self>,
+        m: ProgramNode<NativeField<F>>,
+        remainder_bits: usize,
+    ) -> ProgramNode<Self> {
+        let residues = lhs.residues();
+
+        let mut outputs = vec![];
```

Comment:
optionally `Vec::with_capacity(D)` ; though I'm not sure how important it is to micro-optimize the compilation step.

Reply:
I did some basic profiling and found a way bigger problem than this; we have O(N^2) behavior when traversing the build graph.